signature/src/balrog.rs

```diff
@@ -0,0 +1,382 @@
+    X509(X509Error)
+}
+
+impl From<X509Error> for BalrogError {
+    fn from(e: X509Error) -> Self {
+        BalrogError::X509(e)
+    }
+}
+
+impl From<PEMError> for BalrogError {
+    fn from(e: PEMError) -> Self {
+        BalrogError::PemDecodeError(e.to_string())
+    }
+}
+
+pub fn parse_pem_chain(input: &[u8]) -> Result<Vec<Pem>, PEMError> {
+    Pem::iter_from_buffer(input)
+        .collect::<Result<Vec<Pem>, PEMError>>()
+}
+
+/* A one-and-done function that parses a signature chain and validates a content signature. */
+pub fn parse_and_verify(
+    x5u: &[u8],
+    input: &[u8],
+    signature: &str,
+    current_time: i64,
+    root_hash: &str,
+    leaf_subject: &str
+) -> Result<(), BalrogError> {
+    let pem_chain = match parse_pem_chain(x5u) {
+        Err(e) => return Err(BalrogError::from(e)),
+        Ok(x) => x
+    };
+
+    let balrog = match Balrog::new(&pem_chain) {
+        Err(e) => return Err(BalrogError::from(e)),
+        Ok(x) => x
+    };
+
+    let _ = match balrog.verify_chain(current_time, root_hash) {
+        Err(e) => return Err(BalrogError::from(e)),
+        Ok(x) => x
+    };
+
+    let _ = match balrog.verify_leaf_hostname(leaf_subject) {
+        Err(e) => return Err(e),
+        Ok(x) => x
+    };
+
+    let _ = match balrog.verify_content_signature(input, signature) {
+        Err(e) => return Err(BalrogError::from(e)),
+        Ok(x) => x
+    };
+
+    Ok(())
+}
+
+impl<'a> Balrog<'_> {
+    pub fn new(list: &'a Vec<Pem>) -> Result<Balrog<'a>, X509Error> {
+        if list.is_empty() {
+            return Err(X509Error::InvalidCertificate);
+        }
+
+        let mut parsed: Vec<X509Certificate<'a>> = Vec::new();
+        let mut hash: Vec<u8> = Vec::new();
+        for (i, pem) in list.iter().enumerate() {
+            if pem.label != "CERTIFICATE" {
+                return Err(X509Error::InvalidCertificate);
+            }
+
+            let x509 = match pem.parse_x509() {
+                Err(e) => return Err(X509Error::from(e)),
+                Ok(x) => x
+            };
+            parsed.push(x509);
+
+            /* For the final certificate, calculate its SHA256 hash. */
+            /* TODO: Technically, we can get more bytes in the PEM beyond the end
+             * of the ASN.1 certificate sequence, and this function will include
+             * them in the hash. This feels like an unclear edge case though.
+             */
+            if i+1 == list.len() {
+                let digest = digest::digest(&digest::SHA256, pem.contents.as_slice());
+                hash.extend_from_slice(digest.as_ref());
+            }
+        }
+
+        Ok(Balrog{chain: parsed, root_hash: hash})
+    }
+
+    /* Ensure that the certificate chain is valid. */
+    pub fn verify_chain(
+        &self,
+        current_time: i64,
+        root_hash: &str
+    ) -> Result<(), BalrogError> {
+        assert!(!self.chain.is_empty());
+        assert!(!self.root_hash.is_empty());
```

Comment:
Hm, we should probably handle these cases. This will panic if it is hit. Rust does have `debug_assert` if necessary, although I'd still feel better about handling it.

Reply:
This applies to all subsequent asserts.

Reply:
I'll add handling of these cases. The only way this can occur is if the `Balrog` struct was created manually rather than calling `Balrog::new()` so I felt it was kind of programmer error if we hit these asserts. But it's also not hard to just handle it anyways.